cstructs on heap returned from optional tuple are unpacked with invalid generated C code

**V doctor** 
$ v doctor
```
OS: linux, Ubuntu 22.04.1 LTS
Processor: 16 cpus, 64bit, little endian, AMD Ryzen 7 3700X 8-Core Processor
CC version: cc (Ubuntu 11.2.0-19ubuntu1) 11.2.0

getwd: /home/dominik
vmodules: /home/dominik/.vmodules
vroot: /home/dominik/Projects/vlang
vexe: /home/dominik/Projects/vlang/v
vexe mtime: 2022-10-11 12:11:19
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.1 dab1c8a.34f233c

Git version: git version 2.34.1
Git vroot status: weekly.2022.40-47-g34f233c9 (3 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3
```

**What did you do?**

I've reworked code from issue #16038 to not cause bug but now I triggered another bug related to \[heap\] storage.

```
struct Poss1 {}
struct Poss2 {}

type Possibilities = Poss1 | Poss2

[heap] //comment out this attribute to make error go away BUT now you need to remember to use &PossOwner everytime you create new instance :(
pub struct PossOwner {
    pub:
        name string        
        details Possibilities
}

fn (t PossOwner) get_file(path string) ?(PossOwner, Poss1) {    
	match t.details {
		Poss1 { return t, t.details }
		else { return error("not a file") }
	}
}

fn (item PossOwner) check() ?int {
	assert item.name == "x"
    
	return 0
}

fn somefun() ?int {
	t := PossOwner{name:'x',details:Poss1{}}

	path_info, _ := t.get_file("") or {
		println('NOTOK1')
		return 1
	}

	_ := path_info.check() or {
		println('NOTOK2')
		return 2
	}

	return 0
}

fn main() {
	somefun() or {return }
	println('success')
}

//v -v -cg -keepc -cc gcc vbug5.v 

```

**What did you expect to see?**

`success` on console
 
**What did you see instead?**

```
> C compiler cmd: 'gcc' '@/tmp/v_1000/vbug5.tmp.c.rsp'
73.887   ms C gcc
/tmp/v_1000/vbug5.tmp.c: In function ‘main__somefun’:
/tmp/v_1000/vbug5.tmp.c:187:26: warning: initialization of ‘main__PossOwner **’ from incompatible pointer type ‘multi_return_main__PossOwner_main__Poss1 *’ [-Wincompatible-pointer-types]
  187 | #define HEAP(type, expr) ((type*)memdup((void*)&((type[]){expr}[0]), sizeof(type)))
      |                          ^
/tmp/v_1000/vbug5.tmp.c:12190:43: note: in expansion of macro ‘HEAP’
12190 |         main__PossOwner *(*(path_info)) = HEAP(multi_return_main__PossOwner_main__Poss1, mr_485.arg0);
      |                                           ^~~~
/tmp/v_1000/vbug5.tmp.c:12191:50: error: incompatible type for argument 1 of ‘main__PossOwner_check’
12191 |         _option_int _t3 = main__PossOwner_check((*(path_info)));
      |                                                 ~^~~~~~~~~~~~~
      |                                                  |
      |                                                  main__PossOwner *
/tmp/v_1000/vbug5.tmp.c:12152:51: note: expected ‘main__PossOwner’ but argument is of type ‘main__PossOwner *’
12152 | _option_int main__PossOwner_check(main__PossOwner _v_toheap_item) {
      |                                   ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -529,9 +529,6 @@ fn (mut g Gen) gen_multi_return_assign(node &ast.AssignStmt, return_type ast.Typ
 		styp := if ident.name in g.defer_vars { '' } else { g.typ(node.left_types[i]) }
 		if node.op == .decl_assign {
 			g.write('$styp ')
-			if is_auto_heap {
-				g.write('*')
-			}
 		}
 		if lx.is_auto_deref_var() {
 			g.write('*')

--- a/vlib/v/tests/assign_optional_of_multi_return_with_heap_test.v
+++ b/vlib/v/tests/assign_optional_of_multi_return_with_heap_test.v
@@ -1,0 +1,50 @@
+struct Poss1 {}
+
+struct Poss2 {}
+
+type Possibilities = Poss1 | Poss2
+
+// comment out this attribute to make error go away BUT now you need to remember to use &PossOwner everytime you create new instance :(
+[heap]
+pub struct PossOwner {
+pub:
+	name    string
+	details Possibilities
+}
+
+fn (t PossOwner) get_file(path string) ?(PossOwner, Poss1) {
+	match t.details {
+		Poss1 { return t, t.details }
+		else { return error('not a file') }
+	}
+}
+
+fn (item PossOwner) check() ?int {
+	assert item.name == 'x'
+
+	return 0
+}
+
+fn somefun() ?int {
+	t := PossOwner{
+		name: 'x'
+		details: Poss1{}
+	}
+
+	path_info, _ := t.get_file('') or {
+		println('NOTOK1')
+		return 1
+	}
+
+	_ := path_info.check() or {
+		println('NOTOK2')
+		return 2
+	}
+
+	return 0
+}
+
+fn test_assign_optional_of_multi_return_with_heap() {
+	somefun() or { return }
+	println('success')
+}

